### PR TITLE
contracts and evm process

### PR DESCRIPTION
**Pull request by Tigo9527**

### CODE REVIEW

File: infra/evm/evm.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/sirupsen/logrus"
+	"github.com/tree-graph/bridge-service/helpers"
+	"github.com/tree-graph/bridge-service/infra/contracts/vault"
+	"github.com/tree-graph/bridge-service/infra/database"
+	"github.com/tree-graph/bridge-service/models"
+)
+
+type HandlerEvm struct {
```
[the remark this answers is not in the document]
Comment (wanliqun) [reply infra/evm/evm.go:15]: Besides, it's naming is not very clear. Maybe `EvmClient` or `EvmHandler` would be more appropriate.

File: infra/evm/evm.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	Client *ethclient.Client
+}
+
+var clients map[int64]HandlerEvm
+
+func GetClient(chainId int64) HandlerEvm {
```
Comment (wanliqun) [inline infra/evm/evm.go:21]: renamed as `GetEvmClient` or `GetEvmHandler`?

**Comment by wanliqun** [comment]
please use `gofmt` to format your codes if possible.

File: infra/evm/evm.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	_, _, err := evm.Client.TransactionByHash(context.Background(), common.HexToHash(txHash))
+	if err != nil {
+		return nil, fmt.Errorf("TransactionByHash error: [%v], [%v]", txHash, err.Error())
+	}
+	var bean = models.CrossRequest{ChainId: chainId, Hash: txHash}
+	err = database.DB.Create(&bean).Error
+	if err != nil {
+		return nil, err
+	}
+	logrus.Infof("crossing request added %v %v\n", chainId, txHash)
```
[the remark this answers is not in the document]
Comment (Tigo9527) [reply infra/evm/evm.go]: Changed.

File: infra/evm/evm.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	return &bean, nil
+}
+func (evm *HandlerEvm) ParseCrossChainRequest(hash string) ([]*vault.VaultCrossRequest, error) {
```
[the remark this answers is not in the document]
Comment (Tigo9527) [reply infra/evm/evm.go]: Changed

File: infra/evm/evm.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+			continue
+		}
+		logrus.Infoln("parsed log Asset", parseLog.Asset)
+		logrus.Infoln("parsed log From", parseLog.From)
+		logrus.Infoln("parsed log TargetChain", parseLog.ToChainId)
+		logrus.Infoln("parsed log TokenIds", parseLog.TokenIds)
+		logrus.Infoln("parsed log Amounts", parseLog.Amounts)
+		logrus.Infoln("parsed log Uris", parseLog.Uris)
+		logrus.Infoln("parsed log TargetContract", parseLog.TargetContract)
+		logrus.Infoln("parsed log UserNonce", parseLog.UserNonce)
```
[the remark this answers is not in the document]
Comment (Tigo9527) [reply infra/evm/evm.go]: Changed.

**Review by Tigo9527 — COMMENTED** [review]
Changed

File: infra/evm/evm.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		logrus.Println("create contract fail", err.Error())
+		return nil, err
+	}
+	var requests []*vault.VaultCrossRequest
+	for _, v := range rcpt.Logs {
+		parseLog, err := cf.ParseCrossRequest(*v)
+		if err != nil {
+			logrus.Println("parse log fail", err.Error())
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (Tigo9527) [reply infra/evm/evm.go]: Changed

**Review by wanliqun — APPROVED** [review]

**Pull request merged**